### ISSUE TITLE
Python urllib3 module could it be monkeypatched by eventlet?

### ISSUE THREAD

**Issue by zhang-shengping** [eventlet/eventlet#765]
I use Python `eventlet` pool and `requests` module to speed up my HTTPS request.
But it makes the requests process slower, I try to find the answer to the question these days.

I find some use cases for Python `eventlet`. such as https://stackoverflow.com/questions/5734050/speed-of-fetching-web-pages-with-eventlet-and-python and https://eventlet.net/doc/ these documents mentioned.
```Python
urls = ["http://www.google.com/intl/en_ALL/images/logo.gif",
     "https://wiki.secondlife.com/w/images/secondlife.jpg",
     "http://us.i1.yimg.com/us.yimg.com/i/ww/beta/y3.gif"]

import eventlet
from eventlet.green import urllib2

def fetch(url):

  return urllib2.urlopen(url).read()

pool = eventlet.GreenPool()

for body in pool.imap(fetch, urls):
  print "got body", len(body)
```

``` Python
urls = [
    "http://www.google.com/intl/en_ALL/images/logo.gif",
    "http://python.org/images/python-logo.gif",
    "http://us.i1.yimg.com/us.yimg.com/i/ww/beta/y3.gif",
]

import eventlet
from eventlet.green.urllib.request import urlopen

def fetch(url):
    return urlopen(url).read()

pool = eventlet.GreenPool()
for body in pool.imap(fetch, urls):
    print("got body", len(body))
```

I notice only Python module `urllib` and `urllib2` are used in these examples, and I can not import `urllib3` from `event.green` module.
``` Python
Python 2.7.5 (default, Nov 16 2020, 22:23:17)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from eventlet.green import urllib3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name urllib3
>>>
```

When I check the Python `requests` module, I find it uses `urllib3` module. 

``` Bash
[root@ci-4183549-pzhang requests]# ls
adapters.py   api.py   auth.py   certs.py   compat.py   cookies.py   exceptions.py   hooks.py   __init__.py   models.py   packages      sessions.pyo      status_codes.pyo  structures.pyo  utils.pyo
adapters.pyc  api.pyc  auth.pyc  certs.pyc  compat.pyc  cookies.pyc  exceptions.pyc  hooks.pyc  __init__.pyc  models.pyc  sessions.py   status_codes.py   structures.py     utils.py
adapters.pyo  api.pyo  auth.pyo  certs.pyo  compat.pyo  cookies.pyo  exceptions.pyo  hooks.pyo  __init__.pyo  models.pyo  sessions.pyc  status_codes.pyc  structures.pyc    utils.pyc
[root@ci-4183549-pzhang requests]# ls packages/
chardet  __init__.py  __init__.pyc  __init__.pyo  urllib3
```

The question is: 

Does it mean the `requests` module can not be eventlet monkeypatched, even if I do `eventlet.monkey_patch()` at the beginning of my script? it will not work correctly, will it? 

Any help or hint is appreciated.

**Comment by temoto** [comment]
TL;DR: `import eventlet ; eventlet.monkey_patch()` is enough. It must be the first line executed by your program.

`monkey_patch()` patches stdlib module `socket` used by stdlib `httplib` used by `urllib3` used by `requests`. We ship green version of `socket` module. Rest of modules in this chain do not contain network related code so there is no need to patch or ship special module.

Don't trust me and try it.

```
import eventlet ; eventlet.monkey_patch()
import requests

'urls = ["http://localhost:8080"] * 5

def fetch(url):
    return requests.get(url)

pool = eventlet.GreenPool()
for response in pool.imap(fetch, urls):
        print(f"response {response}")
```

> I use Python eventlet pool and requests module to speed up my HTTPS request.
> But it makes the requests process slower

Greenthread switching is fast but not free. So each request is microseconds slower than on pure Python. But it allows multiple requests to run concurrenly, so N requests as whole execute almost N times faster.

If you worry about speed of each individual request, don't use slow `requests`. Use `urllib3` directly or try `httplib2`.

If you can't guarantee that each url is located on separate domain, you should use connection pool, like `http = requests.Session()` to avoid establishing new connection at each request. **Do this no matter if you use eventlet or not**

If you can, switch to asyncio and `aiohttp` or similar library, then you can have native concurrency from Python itself, no patching.

**Comment by zhang-shengping** [comment]
Thanks for your timely reply @temoto. As you mention using urllib3 directly, I do a performance test by changing the maxsize of `urllib3.PoolManager`. 

I observe two things:

1. every first SSL connection is slow, and the following requests are faster.
2. `maxsize=1` performance is better than `maxsize=1000`. (This is out of my expectations, do you have some idea about this?)

I record the question on (StackOverflow)[https://stackoverflow.com/questions/73064505/python-module-urllib3-increase-the-maxsize-max-pool-size-of-httpsconnectpoo]

**Comment by temoto** [comment]
Hopefully, this is resolved, feel free to reopen.
